### PR TITLE
feat(vd): support filesystem mode

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Support filesystem mode.
Support qcow2 images.

## Why do we need it, and what problem does it solve?
We need to run virtual machines with pvc in file system mode.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [X] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

**Review by LopatinDmitr — DISMISSED** [review]

File: images/virtualization-artifact/pkg/controller/vd/internal/source/http.go [hunk 1]
```diff
@@ -195,8 +195,7 @@ func (ds HTTPDataSource) Sync(ctx context.Context, vd *virtv2.VirtualDisk) (bool
 
 		source := ds.getSource(vd, supgen)
 
-		wffc := vd.Spec.BindingMode != nil && *vd.Spec.BindingMode == virtv2.VirtualDiskBindingModeWaitForFirstConsumer
-		err = ds.diskService.Start(ctx, diskSize, vd.Spec.PersistentVolumeClaim.StorageClass, source, vd, supgen, wffc)
+		err = ds.diskService.Start(ctx, diskSize, vd.Spec.PersistentVolumeClaim.StorageClass, source, vd, supgen)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vd/internal/source/http.go:198]: ErrDefaultStorageClassNotFound is not internal error, write message to condition and return nil

File: images/virtualization-artifact/pkg/controller/service/disk_service.go [hunk 2]
```diff
@@ -66,21 +68,24 @@ func (s DiskService) Start(
...
 	dvBuilder.SetDataSource(source)
-	dvBuilder.SetPVC(storageClass, pvcSize)
 	dvBuilder.SetOwnerRef(obj, obj.GroupVersionKind())
-	dvBuilder.SetBindingMode(wffc)
 
-	// WaitForFirstConsumer is mainly needed for local storage.
-	// To prevent virtual machine migrations from failing, we set PVC to RWO so that virtual machines definitely cannot migrate.
-	if wffc {
-		dvBuilder.SetAccessMode(corev1.ReadWriteOnce)
+	sc, err := s.GetStorageClass(ctx, storageClass)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/disk_service.go:76]: If you only need the name, why not take it from the sprofile?

File: images/virtualization-artifact/pkg/controller/vd/internal/source/http.go (hunk 1 above)
Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/vd/internal/source/http.go:198]: As well as `storageProfile not found`

File: images/virtualization-artifact/pkg/controller/vd/internal/source/upload.go [hunk 3]
```diff
@@ -262,8 +261,11 @@ func (ds UploadDataSource) Sync(ctx context.Context, vd *virtv2.VirtualDisk) (bo
 		if err != nil {
 			return false, err
 		}
-
-		err = setPhaseConditionForPVCProvisioningDisk(ctx, dv, vd, pvc, &condition, ds.diskService)
+		sc, err := ds.diskService.GetStorageClass(ctx, pvc.Spec.StorageClassName)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vd/internal/source/upload.go:264]: Handle not found error: write it to condition and return nil

File: images/virtualization-artifact/pkg/controller/service/disk_service.go [hunk 4]
```diff
@@ -251,6 +256,85 @@ func (s DiskService) GetCapacity(pvc *corev1.PersistentVolumeClaim) string {
...
+	AccessMode corev1.PersistentVolumeAccessMode
+	VolumeMode corev1.PersistentVolumeMode
+}
+
+func (cp StorageCapabilities) IsEmpty() bool {
+	return cp.AccessMode == "" && cp.VolumeMode == ""
+}
+
+func (s DiskService) parseVolumeMode(status cdiv1.StorageProfileStatus) StorageCapabilities {
+	accessModeWeights := map[corev1.PersistentVolumeAccessMode]int{
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/disk_service.go:284]: Take it outside the method: don't create a map every time the method is called

Comment (Isteb4k) [reply images/virtualization-artifact/pkg/controller/service/disk_service.go:284]: Same for scoped functions below

File: images/virtualization-artifact/pkg/controller/service/disk_service.go [hunk 5]
```diff
@@ -251,6 +256,85 @@ func (s DiskService) GetCapacity(pvc *corev1.PersistentVolumeClaim) string {
...
+		for _, mode := range modes {
+			if accessModeWeights[mode] > weight {
+				weight = accessModeWeights[mode]
+				m = mode
+			}
+		}
+		return m
+	}
+
+	var scaps []StorageCapabilities
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/disk_service.go:308]: scaps? The naming is not obvious, better storageCapabilities

File: images/virtualization-artifact/pkg/controller/kvbuilder/dv.go [hunk 6]
```diff
@@ -46,6 +47,7 @@ func NewDV(name types.NamespacedName) *DV {
 					Name:      name.Name,
 					Annotations: map[string]string{
 						"cdi.kubevirt.io/storage.deleteAfterCompletion": "false",
+						cc.AnnUnpackFormat: "qcow2",
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/kvbuilder/dv.go:50]: Move qcow2 to const in imageformat pkg

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/service/disk_service.go (hunk 5 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/service/disk_service.go:308]: scaps: s - storage, caps - capabilities.
So, for me it is obvious, but I will rename it

Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/service/disk_service.go:308]: fixed

File: images/virtualization-artifact/pkg/controller/service/disk_service.go (hunk 2 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/service/disk_service.go:76]: GetStorageClass return default storage class if storageClass was not defined

File: images/cdi-artifact/patches/012-add-caps-for-deckhouse-provisioners.patch [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
++++ b/pkg/storagecapabilities/storagecapabilities.go
+@@ -8,9 +8,10 @@ import (
+ 
+ 	v1 "k8s.io/api/core/v1"
+ 	storagev1 "k8s.io/api/storage/v1"
+-	"kubevirt.io/containerized-data-importer/pkg/util"
+ 	"sigs.k8s.io/controller-runtime/pkg/client"
+ 
++	"kubevirt.io/containerized-data-importer/pkg/util"
++
```
Comment (diafour) [inline images/cdi-artifact/patches/012-add-caps-for-deckhouse-provisioners.patch:13]: This chunk is useless. But ok.

Comment (yaroslavborbat) [reply images/cdi-artifact/patches/012-add-caps-for-deckhouse-provisioners.patch:13]: It's not me, it's my ide.

**Review by Isteb4k — APPROVED** [review]

**Pull request merged**